client/lib/clientEntity.ts

```diff
@@ -0,0 +1,168 @@
+   * @protected
+   */
+  protected _session?: any;
+  /**
+   * @property {NodeJS.Timer} _tokenRenewalTimer The token renewal timer that keeps track of when
+   * the Client Entity is due for token renewal.
+   * @protected
+   */
+  protected _tokenRenewalTimer?: NodeJS.Timer;
+  /**
+   * Creates a new ClientEntity instance.
+   * @constructor
+   * @param {ConnectionContext} context The connection context.
+   * @param {string} [name] Name of the entity.
+   */
+  constructor(context: ConnectionContext, options?: ClientEntityOptions) {
+    if (!options) options = {};
+    this._context = context;
+    this.address = options.address || "";
+    this.audience = options.audience || "";
+    this.name = options.name || uuid();
+    this.partitionId = options.partitionId;
+  }
+  /**
+   * Provides the current type of the ClientEntity.
+   * @return {string} The entity type.
+   */
+  get type(): string {
+    let result = "ClientEntity";
+    if ((this as any).constructor && (this as any).constructor.name) {
+      result = (this as any).constructor.name;
+    }
+    return result;
+  }
+
+  /**
+   * Negotiates the cbs claim for the ClientEntity.
+   * @protected
+   * @param {boolean} [setTokenRenewal] Set the token renewal timer. Default false.
+   * @return {Promise<void>} Promise<void>
+   */
+  protected async _negotiateClaim(setTokenRenewal?: boolean): Promise<void> {
+    // Acquire the lock and establish a cbs session if it does not exist on the connection.
+    // Although node.js is single threaded, we need a locking mechanism to ensure that a
+    // race condition does not happen while creating a shared resource (in this case the
+    // cbs session, since we want to have exactly 1 cbs session per connection).
+    debug("[%s] Acquiring cbs lock: '%s' for creating the cbs session while creating the %s: " +
+      "'%s' with address: '%s'.", this._context.connectionId, this._context.cbsSession!.cbsLock,
+      this.type, this.name, this.address);
+    await defaultLock.acquire(this._context.cbsSession!.cbsLock,
+      () => { return this._context.cbsSession!.init(); });
+    const tokenObject = await this._context.tokenProvider.getToken(this.audience);
+    debug("[%s] %s: calling negotiateClaim for audience '%s'.",
+      this._context.connectionId, this.type, this.audience);
+    // Acquire the lock to negotiate the CBS claim.
+    debug("[%s] Acquiring cbs lock: '%s' for cbs auth for %s: '%s' with address '%s'.",
+      this._context.connectionId, this._context.negotiateClaimLock, this.type, this.name, this.address);
+    await defaultLock.acquire(this._context.negotiateClaimLock, () => {
+      return this._context.cbsSession!.negotiateClaim(this.audience, tokenObject);
+    });
+    debug("[%s] Negotiated claim for %s '%s' with with address: %s",
+      this._context.connectionId, this.type, this.name, this.address);
+    if (setTokenRenewal) {
+      await this._ensureTokenRenewal();
```

Comment:
>this._ensureTokenRenewal(); [](start = 12, length = 27)

Need to be careful with this timer.  What are the paths that clear this timer.  Just dealing with a couple of bugs in the iothub where this timer isn't getting cleared in the case of the client closing.  It then keeps the application from termination.